plots/generic.py

```diff
@@ -0,0 +1,539 @@
+import seaborn as sns
+from elephant.statistics import mean_firing_rate
+from math import log10, floor
+
+
+def _round_to_1(x):
+    rounded = round(x, -int(floor(log10(abs(x)))))
+    return rounded, rounded > x
+
+
+def _get_attributes(spiketrains, key_list):
+    """
+    attribute_array is of shape (len(spiketrains), len(key_list))
+    and consists of numerical ids for each value of each key for each
+    spiketrain.
+
+    Passed spiketrains must be sorted according to key_list.
+    key_list must not be empty.
+    """
+
+    key_count = len(key_list)
+    attribute_array = np.zeros((len(spiketrains), len(key_list)))
+    # count all group sizes for all keys in keylist:
+    while key_count > 0:
+        key_count -= 1
+        group_key = key_list[key_count]
+        i = 0
+        if group_key in spiketrains[i].annotations:
+            current_value = spiketrains[i].annotations[group_key]
+        else:
+            # use placeholder value when key is not in annotations
+            # of the current spiketrain
+            current_value = '####BLANK####'
+        ref_value = current_value
+        values = np.array([])
+        # count all group sizes for values of current key:
+        while i < spiketrains.__len__():
+            if not len(values) or current_value not in values:
+                values = np.append(values, current_value)
+            # count group size for a value of the current key:
+            while i < len(spiketrains) and current_value == ref_value:
+                attribute_array[i][key_count] = \
+                np.where(values == current_value)[0][0]
+                i += 1
+                if i < len(spiketrains):
+                    if group_key in spiketrains[i].annotations:
+                        current_value = spiketrains[i].annotations[
+                            group_key]
+                    else:
+                        current_value = '####BLANK####'
+            ref_value = current_value
+    return attribute_array
+
+
+def rasterplot(spiketrain_list,
+               key_list=[],
+               groupingdepth=0,
+               spacing=[5,3],
+               colorkey=0,
+               pophist_mode='color',
+               pophistbins=100,
+               right_histogram=mean_firing_rate,
+               filter_function=None,
+               histscale=.1,
+               labelkey=None,
+               markerargs={'markersize':4,'marker':'.'},
+               separatorargs=[{'linewidth':2, 'linestyle':'--', 'color':'0.8'},
+                              {'linewidth':1, 'linestyle':'--', 'color':'0.8'}],
+               legend=False,
+               legendargs={'loc':(.98,1.), 'markerscale':1.5, 'handletextpad':0},
+               ax=None,
+               style='ticks',
+               palette='Set2',
+               context='paper',
+               colorcodes='colorblind'):
+
+    """
+    This function plots the dot display of spiketrains alongside its population
```

Comment:
Please reduce to a single sentence as first paragraph of the documentation to say roughly what this function does (e.g., only: This function plots the dot display of spiketrains alongside its population histogram and the mean firing rate (or a custom function).)

Then in a second paragraph be a bit more specific about what this function does.